# Call Spots

The Call Reference Spots section of the pipeline is a method of gene calling which runs quickly on a small set of spots ($\approx$ 50, 000 per tile) of the anchor image. Initially, this was our final mode of gene calling, but has since been superseded by OMP, which differs from Call Spots in that it runs on several more pixels, regardless of whether they have been detected as a spot.

Despite this, the call spots section is still a crucial part of the pipeline as it estimates several important parameters used in the OMP section.

Some of the most important exported parameters of this section are:

- **Colour Normalisation Factor** $\mathbf{A}$:  $(n_{\text{t}} \times n_{\text{r}} \times n_{\text{c}})$ array which multiplies the colours to minimise any systematic brightness variability between different tiles, rounds and channels and maximise spectral separation of dyes,

- **Bleed Matrix** $\mathbf{B}$: $(n_{\text{d}} \times n_{\text{c}})$ array of the typical channel spectrum of each dye.

- **Bled Codes** $\mathbf{K}$: $(n_{\text{g}} \times n_{\text{r}} \times n_{\text{c}})$ array of the expected colour spectrum for each gene.

## Algorithm Breakdown

<figure markdown="span">
  ![Image title](images/algorithm/call_spots/Call Spots Flowchart.png)
    <span> Algorithm Flowchart showing how all variables and steps of the pipeline are related.</span>
</figure>


The inputs to the algorithm are:

- Raw spot colours $F_{src}$ for all spots $s$ (defined as local maxima of round $r_{\text{anchor}}$, channel $c_{\text{anchor}}$) and the tile $t(s)$ they belong to.

- A list of genes $g$ and their associated dye codes $d(g, r)$ for each round $r$. These codes were generated by the Reed-Solomon Algorithm which should minimise the number of overlap between codes.
 
- A raw bleed matrix $\mathbf{B_{\textrm{raw}}}$ of shape $(n_{\text{dyes}} \times n_{\text{c}})$ obtained from images of free-floating drops of each dye.

### 0: Preprocessing
The purpose of this step is to: 

- Approximately equalise the brightness of different tiles, rounds and channels,
- Remove any persistent background signal in each channel from each spot colour.

We transform the raw spot colours $F_{src}$ as follows:

$$F_{src} \mapsto \tilde{A}_{t(s)rc}F_{src} - G_{sc}.$$ 

In the formula above:

- The _initial normalisation factor_ $\tilde{A}_{trc}$ is defined as

$$
\tilde{A}_{trc} = \dfrac{1}{\text{Percentile}_s(F_{src}, 95)}
$$

for all spots $s$ in tile $t$. This is a good estimate of the scaling factor needed to make the brightest spots in each tile, round and channel have the same intensity.

- The _ground level_ $G_{sc}$ is defined as

$$
G_{sc} = \text{Percentile}_r(\tilde{A}_{t(s)rc}F_{src}, 25).
$$

For 7 rounds, this is the brightness of the second dimmest round of the scaled spot colours in channel $c$. This is a good estimate of the constant signal in channel $c$ across all rounds, which we want to remove.


<!-- TODO: Add an image of a spot before anything, after scaling, then after removing background -->

### 1: Initial Gene Assignment
The purpose of this step is to provide some preliminary gene assignments that will allow us to estimate the bleed matrix and the bled codes. We will work extensively with the bleed matrix in these calculations, but bear in mind that this is the raw bleed matrix $\mathbf{B_{\textrm{raw}}}$.

#### Gene Probability Calculation

We'd like to define a probability that each spot $s$ belongs to each gene $g$. We'd like this probability to have the following properties:

1. The probability of each round $r$ being assigned to a particular dye $d$ should be independent of overall brightness of the spot in that round, ie: the norm of the round colour vector $\mathbf{F_{sr}}$,

2. The probabilities of each round $r$ should be independent of each other.

??? note "Why these properties?"
    
    Property 1 is desirable because of the way the bridge probes work in the experiment, as shown below. 

    <p align="center">
      <img src="https://github.com/user-attachments/assets/e452ca8d-cf7d-4e64-8203-737079aaa11c" width="600" />
      <br />
      <span> The 3 probe system.</span>
    </p>

    Upon every mRNA transcript of interest, several padlock-probes are attached. These stay fixed in place throughout the experiment. To illuminate each gene with the expected dye in each round, bridge probes (which are gene-specific on one arm and dye-specific on the other) are transported to all the padlock probes associated with this spot and ligated.

    The brightness of each gene in each round is proportional to the amount of bridge probes that have ligated. Unfortunately, the number of bridge probes varies wildly between genes and rounds, giving rise to systematic differences in brightness between genes and rounds. 
    
    Normalising the brightnesses of each spot in each round is a way to get around this problem.

    <p align="center">
      <img src="https://github.com/user-attachments/assets/be170d2d-9dc1-4bb1-bfaa-eaca90a3d0a8" width="600" />
      <br />
      <span> Spots assigned to the gene Chrm 1 had many more bridge probes in round 4 than round 3, leading to systematic brightness differences.</span>
    </p>
    
    Property 2 is only approximately true, but independence between rounds makes the formula for the probability of a spot being assigned to a gene much simpler.

Because of property 1, we only need to define a probability on unit vectors, ie: vectors $\mathbf{f_r}$ with $\|\mathbf{f_r}\| = 1$. Because of property 2, it is sufficient to define a probability of each round $r$ being assigned to a particular dye $d$.

##### Probabilistic Dye Assignments

We'll model the normalised round fluorescence vectors $\mathbf{F_r}$ arising from dye $d$ as being random and distributed according to a von Mises-Fisher distribution with mean $\mathbf{B_d}$ and concentration parameter $\kappa$.

This model has probability density function

$$\mathbb{P}[\mathbf{F_{r}} = \mathbf{f_r} \mid D = d] =  M_{\kappa} \exp(\kappa\mathbf{f_r} \cdot \mathbf{B_d}),$$

where $\mathbf{f_r}$ is a unit vector and $M_{\kappa}$ is a normalization constant we donâ€™t need to worry about.

##### Probabilistic Gene Assignments

Now let $\mathbf{F} = (\mathbf{F_1}, \ldots, \mathbf{F_{n_r}}) ^ T$ be the $(n_r \times n_c)$ matrix of normalised fluorescence vectors of each round $r$ of a spot $s$. We'd like to find the probability that a spot with fluorescence $\mathbf{F}$ is assigned to gene $g$. Independence of rounds means this is the product of the probabilities that each round $r$ is assigned to dye $d(g, r)$.

$$ 
\begin{aligned}
\mathbb{P}[\mathbf{F} = \mathbf{f} \mid G = g] &= \prod_r \mathbb{P}[\mathbf{F_{r}} = \mathbf{f_r} \mid D = d(g, r)] \\ 
&= \prod_r M_{\kappa} \exp \left( \kappa\mathbf{f_r} \cdot \mathbf{B_{d(g, r)}} \right) \\
&= M_{\kappa}^{n_r} \exp \left( \kappa \sum_r \mathbf{f_r} \cdot \mathbf{B_{d(g, r)}} \right) \\ 
&=  M_{\kappa}^{n_r} \exp(\kappa \mathbf{f \cdot b_g}), \end{aligned}
$$

where

- $\mathbf{f} = (\mathbf{f_1}, \ldots, \mathbf{f_{n_r}}) ^ T$ is the observed round-normalised $(n_r \times n_c)$ fluorescence matrix of the spot,

- $\mathbf{b_g} = (\mathbf{B_{d(g, 1)}}, \ldots, \mathbf{B_{d(g, n_r)}}) ^ T$ is the $(n_r \times n_c)$ matrix of the bled code for gene $g$,

- the dot product $\mathbf{f \cdot b_g}$ is the Frobenius inner product, ie: the sum of the elementwise products of the two matrices.

We have so far only defined the probability of $\mathbf{F} = \mathbf{f}$ given $G = g$. We can find the probability of $G = g$ given $\mathbf{F} = \mathbf{f}$ using Bayes' Theorem:

$$ 
\mathbb{P}[G = g \mid \mathbf{F} = \mathbf{f}] 
= \dfrac{\mathbb{P}[\mathbf{F} = \mathbf{f} \mid G = g] \mathbb{P}[G = g]}{ \mathbb{P}[\mathbf{F} = \mathbf{f}]}. 
$$

For the priors, we will assume that:

- $\mathbb{P}[G = g] = \frac{1}{n_g}$, so all genes are equally likely before we see any data,

- $\mathbb{P}[\mathbf{F} = \mathbf{f}] = \sum_g \mathbb{P}[\mathbf{F} = \mathbf{f} \mid G = g] \mathbb{P}[G = g] = \frac{1}{n_g} \sum_g M_{\kappa}^{n_r} \exp(\kappa \mathbf{b}_g \cdot \mathbf{f})$, which is implicitly assuming $\mathbf{F}$ definitely comes from 1 of the genes.

This gives us the final probability:

$$ \mathbb{P}[G = g \mid \mathbf{F} = \mathbf{f}] = \frac{\exp(\kappa \mathbf{b}_g \cdot \mathbf{f})}{\sum_g \exp(\kappa\mathbf{b}_g \cdot \mathbf{f} )}$$

??? note "How to choose $\kappa$?"

    The parameter $\kappa$ is set by adjusting the config parameter `kappa` and has default value 2. The value of $\kappa$ controls how much the probabilities are spread out among the genes, but does not influence the gene ordering.
    
    - $\kappa = 0$ yields a uniform distribution of probabilities between all genes,
    
    - $\kappa \rightarrow \infty$ yields a ditribution that tends to 1 for the gene with the maximum dot product and 0 for all others.

    When working with larger gene panels, all probabilities are spread out more naturally, so it helps to increase $\kappa$ so that probabilities have a consistent interpretation. Out current implementation sets $\kappa = 2$ if $n_g < 200$ and 3 otherwise.


This shows that the value of $\kappa$ has no effect on the gene ordering, but just on the spread of probabilities among genes. A value of 0 yields a uniform distribution of probabilities between all genes, while a very large value of $\kappa$ is approximately 1 for the gene with the maximum dot product and 0 for all others.

### 2: Bleed Matrix Calculation
The purpose of this step is to compute an updated estimate of the bleed matrix. To get a more accurate estimate of the colour spectrum for each dye $d$ we will find high probability spots for genes $g$ containing dye $d$ and then extract the rounds in which dye $d$ is supposed to be present. We will then use singular-value decomposition to get a representative vector for the set of samples. 

Set some probability threshold $\alpha$ (by default, $\alpha = 0.9$). We define the following sets:

$$ 
\mathcal{S} = \{ s : p(s) \geq \alpha \},
$$

$$ 
G_{dr} = \{ g : d(g,r) = d \},
$$

$$ 
J_{dr} = \{ \mathbf{F_{sr}} : s \in \mathcal{S}, \ g_s \in G_{dr} \}
$$

where: 

- $\mathcal{S}$ is the set of spots with $s$ with probability $p(s) > \alpha$,

- $G_{dr}$ is the set of genes with dye $d$ in round $r$,

- $J_{dr}$ is the set of colours of high probability spots in round $r$, for all spots assigned to a gene $g \in G_{dr}$.

By taking the union of $J_{dr}$ across rounds, we end up with a set of reliable colour vector estimates for dye $d$:

$$ 
\mathcal{J}_d = \bigcup_r J_{dr}
$$

Let $\mathbf{J}$ be the $(n_{\text{good spots}} \times n_c)$ matrix form of the set $\mathcal{J}_d$. This just means each row of $\mathbf{J}$ corresponds to a good spot and each column corresponds to a channel.

We'd like to find a representative colour of the matrix $\mathbf{J}$. We do this by computing the first singular vectors of $\mathbf{J}$. These are 2 unit vectors:

- $\boldsymbol{\eta}$ (length $n_{\text{good spots}}$)
- $\boldsymbol{\omega}$ (length $n_c$)

such that 

$$ J_{s, c} \approx \lambda \eta_s \omega_c, $$

for some $\lambda$. We then set the bleed matrix for dye $d$ to $\boldsymbol{\omega}$,  ie: $\mathbf{B_d} = \boldsymbol{\omega}$, which is a normalised fluorescence vector for dye $d$.

### 3: Free Bled Code Estimation

**Terminology:**

- A _free bled code_ $\mathbf{b_g}$ is a bled code computed by an averaging process among spots assigned to gene $g$. This bled code $g$ is _free_ in the sense that the only thing influencing it is the spots of gene $g$.

- A _constrained bled code_, (usually referred to as just a _bled code_) $\mathbf{b_g}$ for a gene $g$ is a bled code which is influenced by parameters calculated across many genes. This is _constrained_ in the sense that parameters other than gene $g$ alone influence our estimate of $\mathbf{b_g}$.

The purpose of this step is to compute normalised free bled codes $\mathbf{b_g}$ for each gene $g$. We will make extensive use of the initial gene assignments and the bleed matrix we have computed, which forms the prior estimate of any fluorescence vector estimate. We will estimate tile-dependent free bled codes, $\mathbf{D_{gtrc}}$ and tile-independent free bled codes, $\mathbf{E_{grc}}$. By comparing these, we will be able to correct for tile by tile variations in brightness in step 5.

Our method of estimating $\mathbf{E_{g,r}}$ (and $\mathbf{D_{g,t, r}}$ just on a smaller set of sample spots):

- Assumes a prior of $\mathbf{E_{g,r}} = \mathbf{B_{d(g,r)}}$,

- Easily scales $\mathbf{B_{d(g,r)}}$ but needs many spots to change its direction,

- Allows some change in the direction if there are sufficiently many spots, meaning that we allow different genes to show the same dye differently if we have enough evidence this is real. This is particularly relevant when dyes fail to completely wash out between rounds.

<figure markdown="span">
  ![Image title](images/algorithm/call_spots/bayes_mean_single.png)
</figure>
*The Bayes Mean biases the sample mean towards a prior vector. This is useful when the number of samples is small and we expect the points to have mean parallel to the prior vector.*


#### The Parallel Biased Bayes Mean

Fix a gene $g$ and round $r$ and let $\mathbf{F_{1,r}}, \ldots, \mathbf{F_{n,r}}$  be the round $r$ fluorescence vectors of spots assigned to gene $g$ with high probability.

We'd like to find a representative vector for this data which captures the mean length, but some genes have very few spots so an average is noisy. We therefore bias our mean towards a scaled version of $\mathbf{B}_{d(g,r)}$. 

To begin, assume the vectors $\mathbf{F_{1,r}}, \ldots, \mathbf{F_{n,r}}$ are i.i.d normal random variables with mean $\boldsymbol{\mu}$ and covariance $I_{n_c}$. Impose a normal prior on the space of possible means:

$$
\overline{\mathbf{F}} \sim \mathcal{N} \bigg( \boldsymbol{\mu}, \frac{\boldsymbol{I_{n_c}}}{n} \bigg)
$$

$$
\boldsymbol{\mu} \sim \mathcal{N}(\mathbf{B}_{d(g,r)}, \Sigma)
$$

where 

$$
\Sigma = \text{Diag}\left(\frac{1}{\alpha^2}, \frac{1}{\beta^2}, \ldots, \frac{1}{\beta^2}\right),
$$

in the orthonormal basis $\mathbf{v}_1 = \mathbf{B}_{d(g,r)}$, and everything else orthogonal to this.

The parameters $\alpha$ and $\beta$, being inverse variances, are concentration parameters. We'll let $\alpha << \beta$, so the set of probable means is very loosely concentrated along the line $\lambda \mathbf{B}_{d(g,r) }$, but very tightly concentrated perpendicular to this line!

Set $\boldsymbol{\Lambda} =\boldsymbol{\Sigma}^{-1}$, $\mathbf{b} = \mathbf{B_{d(g,r)}}$, and recall that the normal is a conjugate prior, ie: the posterior $\boldsymbol{\mu} \mid \mathbf{\overline{F}}$ is normal, so its mean is its mode. 

To find its mode we'll solve for the zeros of the derivative of its log-density. The log-density of $\boldsymbol{\mu} \mid \mathbf{\overline{F}}$ is given by

$$
\begin{aligned}
l(\boldsymbol{\mu}) &= \log P(\boldsymbol{\mu}| \overline{\mathbf{F}} = \mathbf{f}) \\ \\
       &= \log P(\boldsymbol{\mu}) + \log P(\overline{\mathbf{F}} = \mathbf{f} | \boldsymbol{\mu}) + C \\ \\
       &= -\frac{1}{2} (\boldsymbol{\mu} - \mathbf{b})^T \boldsymbol{\Lambda} (\boldsymbol{\mu} - \mathbf{b}) - \frac{n}{2} (\boldsymbol{\mu} - \mathbf{f})^T (\boldsymbol{\mu} - \mathbf{f}) + C
\end{aligned}
$$

This has derivative

$$
\frac{\partial l}{\partial \boldsymbol{\mu}} = - \boldsymbol{\Lambda} (\boldsymbol{\mu} - \boldsymbol{b}) - n(\boldsymbol{\mu} - \mathbf{f})
$$

Setting this to $\mathbf{0}$, rearranging for $\boldsymbol{\mu}$ and using the fact that

$$
\boldsymbol{\Lambda} \mathbf{v} = 
\begin{cases}
\alpha^2 \mathbf{v} & \text{if } \mathbf{v} = \lambda\mathbf{b} \\ \\
\beta^2 \mathbf{v} & \text{otherwise}
\end{cases}
$$

we get

$$
\begin{aligned}
\mathbf{\hat{\mu}} &= (\Lambda + nI)^{-1}(\Lambda \mathbf{b} + n\mathbf{f}) \\ \\ 
    &= (\Lambda + nI)^{-1}(\alpha^2 \mathbf{b} + n\mathbf{f}) \\ \\
    &= (\Lambda + nI)^{-1}(\alpha^2 \mathbf{b} + n(\mathbf{f} \cdot \mathbf{b})\mathbf{b} + n(\mathbf{f} - (\mathbf{f} \cdot \mathbf{b})\mathbf{b})) \\ \\
    &= (\Lambda + nI)^{-1}((\alpha^2 + n \mathbf{f} \cdot \mathbf{b})\mathbf{b} + n(\mathbf{f} - (\mathbf{f} \cdot \mathbf{b})\mathbf{b}))\\ \\
&= \dfrac{(\alpha^2 + n \mathbf{f} \cdot \mathbf{b})}{n + \alpha^2} \mathbf{b} +
 \dfrac{n}{n+\beta^2} \bigg( \mathbf{f} - (\mathbf{f} \cdot \mathbf{b})\mathbf{b} \bigg)
\end{aligned}
$$

Plugging in $\mathbf{f} = \frac{1}{n}\sum_i \mathbf{F_{i, r}}$ yields our estimate $\mathbf{\hat{\mu}}$

<figure markdown="span">
  ![Image title](images/algorithm/call_spots/Bayes Mean.png)
  </figure>
*Decreasing $\beta$ increases the component of the Bayes Mean $\boldsymbol{\hat{\mu}}$ perpendicular to the prior vector. The values of $\alpha^2$ and $\beta^2$ should be thought of as the number of spots needed to change the scale and direction respectively of the prior vector.*

### 4: Round and Channel Normalisation

The purpose of this step is to try and remove systematic brightness differences between rounds and channels and ensure that the dyes are well separated. 

To begin, fix a round $r$ and channel $c$ and let $d_{max}(c)$ be the dye which is most intense in channel $c$. We define a target value $T_d$ for each dye $d$ in its maximal channel $c_{max}(d)$. Now let $G_{r,d}$ be the set of genes with dye $d$ in round $r$, and define the loss function 

$$
L(V_{r, c}) = \sum_{g \in G_{r, \ d_{max}(c)}} \sqrt{N_{g}} \  \bigg( V_{r, c} \ E_{grc} - T_{d_{max}(c)} \bigg)^2,
$$ 

where $N_g$ is the number of high probability spots assigned to gene $g$. There is no reason this has to be a square root, though if it is not, too much influence is given to the most frequent genes. Minimise this loss to obtain

$$
V_{r, c} = \dfrac{ \sum_{g \in G_{r, \ d_{max}(c) }} \sqrt{N_g} E_{grc} T_{d_{max}(c)} } { \sum_{g \in G_{r, \ d_{max}(c) }} \sqrt{N_g} E_{grc}^2 },
$$

which is our optimal value!


<figure markdown="span">
  ![Image title](images/algorithm/call_spots/rc scale regression.png)
</figure>
*The gene intensities for each round and channel plotted in cyan, and their scaled versions plotted in red, showing how they have been recentred around the target values. Plot generated using the `view_rc_scale_regression` diagnostic*

<figure markdown="span">
  ![Image title](images/algorithm/call_spots/rc_scales_im.png)
</figure>
*The round/channel scale matrix shows that most of its job is boosting channel 15 in this case, but the amount it boosts these values is highly variable between rounds.*

Now define the _constrained bled codes_, which we will just call _bled codes_ 

$$
K_{grc} = E_{grc}V_{rc}.
$$

We will use these instead of $E_{grc}$ from here onwards.

<figure markdown="span">
  ![Image title](images/algorithm/call_spots/free vs constrained bled codes.png)
</figure>
*The `view_free_and_constrained_bled_codes` diagnostic shows the free tile-independent bled codes next to the constrained bled codes. The constrained bled codes have much more uniform colours within each gene. Note how the final round and channel has been scaled down and all of channel 15 has been scaled up.*
### 5: Tile Normalisation

The purpose of this step is to remove brightness differences between tiles, and improve the round and channel normalisation we found in the previous step. We do this by finding a scale factor $Q_{trc}$ such that 

$$
Q_{trc} D_{gtrc} \approx K_{grc},
$$

where $\mathbf{D_{g, t,}}$ is the tile-dependent bled code for gene $g$ in tile $t$ defined in step 3 and $\mathbf{K_g}$ is the constrained bled code for gene $g$ defined in step 4.

Our method works in a similar way to step 4: fix a tile $t$, round $r$ and channel $c$ and as above, let $G_{r,d}$ be the genes with dye $d$ in round $r$. Define the loss

$$
L(Q_{trc}) = \sum_{g \in G_{r, \ d_{max}(c)}} \sqrt{N_{g,t}} \  \bigg( Q_{trc} \ D_{gtrc} - K_{grc} \bigg)^2,
$$ 

where $N_{g, t}$ is the number of high probability spots of gene $g$ in tile $t$. Remember that $K_{grc} = E_{grc}V_{r, c},$ so writing this in full yields

$$
L(Q_{trc}) = \sum_{g \in G_{r, \ d_{max}(c)}} \sqrt{N_{gt}} \  \bigg( Q_{trc} \ D_{gtrc} - E_{grc}V_{rc} \bigg)^2.
$$ 

This means that if $D_{gtrc} \approx E_{grc}$ for all genes $g$ then $Q_{t,r,c} \approx V_{r,c}$. This means that $\mathbf{Q}$ is correcting for tile differences. Then why does it have indices for $r$ and $c$? Because the way that the brightness varies between tiles may be completely independent for different round-channel pairs. This is addressed further in the diagnostics. Minimising this loss yields:

$$
Q_{trc} = \dfrac{ \sum_{g \in G_{r, \ d_{max}(c) }} \sqrt{N_{gt}} \ K_{grc}  D_{gtrc}} { \sum_{g \in G_{r,\ d_{max}(c) }} \sqrt{N_{gt}}  D_{gtrc}^2 }.
$$

<figure markdown="span">
  ![Image title](images/algorithm/call_spots/tile scale regression.png)
</figure>
*We can use the `view_tile_scale_regression` diagnostic to view the regression for a single tile. Note that the regressions seem to have a high $r^2$ value and the slopes are significantly different even within channels.*

<figure markdown="span">
  ![Image title](images/algorithm/call_spots/different scales.png)
</figure>
*We can use the `view_scale_factors` diagnostic to view the round/channel scale, the tile scale and the tile scales relative to the round/channel scales. The final plot being non-uniform implies that the tile scale corrections are independent between different $(r, c)$ pairs.*

### 6 and 7: Application of Scales, Computation of Final Scores and Bleed Matrix

The purpose of this step is to bring all the components together and compute the final scores and bleed matrix.

Now that we have the tile scale $Q_{t,r,c}$, we multiply it by the initial scale factor $P_{trc}$ to get our final colour normalisation factor $A_{trc}$:

$$
A_{trc} = P_{trc} Q_{trc}.
$$

This is important as all our calculations have been done on preprocessed spot colours which have already been multiplied by $\mathbf{P}$. We apply this scale to all of our spot colours $F$ by pointwise multiplication.

Next, we compute the final probabilities and dot product scores. We might ask at this point whether we should use:

1. The tile-independent free bled codes $E_{grc}$, 

2. the tile-dependent free bled codes $D_{gtrc}$ or,

3. the constrained bled codes $K_{grc}$? 

The best answer is definitely the constrained bled codes $K_{grc}$! We calculated $\mathbf{E}$ and $\mathbf{D}$ as important summary statistics which act as representative samples for each gene in the case of $\mathbf{E}$, and for each gene and tile in the case of $\mathbf{D}$. These were only computed to facilitate the computations of $\mathbf{A}$ and $\mathbf{K}$.

While it may seem more accurate to have a different gene code for each tile, the estimates $\mathbf{D_{g, t}}$ are noisy due to small numbers of samples. The colour normalisation factor $\mathbf{A}$ was calculated specifically to maximise similarity of the **tile-dependent** free codes $\mathbf{D}$ with the **tile-independent** constrained codes $\mathbf{K}$, meaning that multiplying spot colours by $\mathbf{A}$ has the dual effect of

1. homogenising them across tiles and,

2. bringing them all close to the constrained codes $\mathbf{K}$.

With that in mind, we compute:

1. Final gene probabilities using the scaled spots $\mathbf{AF}$ and comparing against the constrained codes $\mathbf{K}$

2. Final Dot Products using the scaled spots $\mathbf{AF}$ and comparing against the constrained codes $\mathbf{K}$. These would not have been accurate in step 1 as we had no model of how each gene varied in brightness between rounds, but now this is something we have accounted for in $\mathbf{K}$.

3. The Final Bleed Matrix using the same method as discussed in step 2, but with updated gene probabilities.
