# Call Spots

The Call Reference Spots section of the pipeline is a method of gene calling which runs quickly on a small set of spots ($\approx$ 50, 000 per tile) of the anchor image. Initially, this was our final mode of gene calling, but has since been superseded by OMP, which differs from Call Spots in that it runs on several more pixels, regardless of whether they have been detected as a spot.

Despite this, the call spots section is still a crucial part of the pipeline as it estimates several important parameters used in the OMP section.

Some of the most important exported parameters of this section are:

- **Colour Normalisation Factor** $\mathbf{A}$:  $(n_{\text{t}} \times n_{\text{r}} \times n_{\text{c}})$ array which multiplies the colours to minimise any systematic brightness variability between different tiles, rounds and channels and maximise spectral separation of dyes,

- **Bleed Matrix** $\mathbf{B}$: $(n_{\text{d}} \times n_{\text{c}})$ array of the typical channel spectrum of each dye.

- **Bled Codes** $\mathbf{K}$: $(n_{\text{g}} \times n_{\text{r}} \times n_{\text{c}})$ array of the expected colour spectrum for each gene.

## Algorithm Breakdown

<figure markdown="span">
  ![Image title](images/algorithm/call_spots/Call Spots Flowchart.png)
    <span> Algorithm Flowchart showing how all variables and steps of the pipeline are related.</span>
</figure>


The inputs to the algorithm are:

- Raw spot colours $F_{src}$ for all spots $s$ (defined as local maxima of round $r_{\text{anchor}}$, channel $c_{\text{anchor}}$) and the tile $t(s)$ they belong to.

- A list of genes $g$ and their associated dye codes $d(g, r)$ for each round $r$. These codes were generated by the Reed-Solomon Algorithm which should minimise the number of overlap between codes.
 
- A raw bleed matrix $\mathbf{B_{\textrm{raw}}}$ of shape $(n_{\text{dyes}} \times n_{\text{c}})$ obtained from images of free-floating drops of each dye.

### 0: Preprocessing
The purpose of this step is to approximately equalise the brightness of different tiles, rounds and channels and to remove any background which is constant across rounds from each spot.

We transform the raw spot colours $F_{src}$ as follows:

$$F_{src} \mapsto \tilde{A}_{t(s)rc}F_{src} - G_{sc}.$$ 

In the formula above:

- The _initial normalisation factor_ $\tilde{A}_{trc}$ is defined as

$$
\tilde{A}_{trc} = \dfrac{1}{\text{Percentile}_s(F_{src}, 95)}
$$

for all spots $s$ in tile $t$. This is a good estimate of the scaling factor needed to make the brightest spots in each tile, round and channel have the same intensity.

- The _ground level_ $G_{sc}$ is defined as

$$
G_{sc} = \text{Percentile}_r(\tilde{A}_{t(s)rc}F_{src}, 25).
$$

For 7 rounds, this is the brightness of the second dimmest round of the scaled spot colours in channel $c$. This is a good estimate of the constant signal in channel $c$ across all rounds, which we want to remove.


<!-- TODO: Add an image of a spot before anything, after scaling, then after removing background -->

### 1: Initial Gene Assignment
The purpose of this step is to provide some preliminary gene assignments that will allow us to estimate the bleed matrix and the bled codes. We will work extensively with the bleed matrix in these calculations, but bear in mind that this is the raw bleed matrix $\mathbf{B_{\textrm{raw}}}$.

We'd like to define a probability that spot $s$ (fluorescence $F_{src}$) comes from gene $g$, and we'd like this probability to have the following properties:

1. The probability of round $r$ being assigned to dye $d$ should be invariant to changes in the overall brightness of $\mathbf{F_{sr}}$,

2. the probabilities of each round $r$ should be independent.


??? note "Why these properties?"
    
    Property 1 is desirable because of the way the bridge probes work in the experiment, as shown below. 

    <p align="center">
      <img src="https://github.com/user-attachments/assets/e452ca8d-cf7d-4e64-8203-737079aaa11c" width="600" />
      <br />
      <span> The 3 probe system.</span>
    </p>

    Upon every mRNA transcript of interest, several padlock-probes are attached. These stay fixed in place throughout the experiment. To illuminate each gene with the expected dye in each round, bridge probes (which are gene-specific on one arm and dye-specific on the other) are transported to all the padlock probes associated with this spot and ligated.

    The brightness of each gene in each round is proportional to the amount of bridge probes that have ligated. Unfortunately, the number of bridge probes varies wildly between genes and rounds, giving rise to systematic differences in brightness between genes and rounds. 
    
    Normalising the brightnesses of each spot in each round is a way to get around this problem.

    <p align="center">
      <img src="https://github.com/user-attachments/assets/be170d2d-9dc1-4bb1-bfaa-eaca90a3d0a8" width="600" />
      <br />
      <span> Spots assigned to the gene Chrm 1 had many more bridge probes in round 4 than round 3, leading to systematic brightness differences.</span>
    </p>
    
    Property 2 is only approximately true, but independence between rounds makes the formula for the probability of a spot being assigned to a gene much simpler.

Let:

- $\mathbf{f} = (\mathbf{f_1}, \ldots, \mathbf{f_{n_r}}) ^ T$ be the $(n_r \times n_c)$ round-normalised fluorescence matrix of the spot,

- $\mathbf{b}_g = (\mathbf{B_{d(g, 1)}}, \ldots, \mathbf{B_{d(g, n_r)}}) ^ T$ be the $(n_r \times n_c)$ round-normalised bled code for gene $g$.

We define the probability of spot $s$ being assigned to gene $g$ as

$$ \mathbb{P}[G = g \mid \mathbf{F} = \mathbf{f}] = \frac{\exp(\kappa \mathbf{b}_g \cdot \mathbf{f})}{\sum_g \exp(\kappa\mathbf{b}_g \cdot \mathbf{f})},$$

where $\kappa$ is a concentration parameter which controls how much the probabilities are spread out among the genes.

!!! tip "How to choose $\kappa$?"

    The parameter $\kappa$ is set by adjusting the config parameter `kappa` and has default value 2. The value of $\kappa$ controls how much the probabilities are spread out among the genes, but does not influence the gene ordering.
    
    - $\kappa = 0$ yields a uniform distribution of probabilities between all genes,
    
    - $\kappa \rightarrow \infty$ yields a ditribution that tends to 1 for the gene with the maximum dot product and 0 for all others.

    When working with larger gene panels, all probabilities are spread out more naturally, so it helps to increase $\kappa$ so that probabilities have a consistent interpretation. Out current implementation sets $\kappa = 2$ if $n_g < 200$ and 3 otherwise.


??? info "Gene Probability Derivation"

    #### Dye Probabilities
    
    We'll model the normalised round fluorescence vectors $\mathbf{F_r}$ arising from dye $d$ as being random and distributed according to a von Mises-Fisher distribution with mean $\mathbf{B_d}$ and concentration parameter $\kappa$.
    
    This model has probability density function
    
    $$\mathbb{P}[\mathbf{F_{r}} = \mathbf{f_r} \mid D = d] =  M_{\kappa} \exp(\kappa\mathbf{f_r} \cdot \mathbf{B_d}),$$
    
    where $\mathbf{f_r}$ is a unit vector and $M_{\kappa}$ is a normalization constant we donâ€™t need to worry about.
    
    #### Gene Probabilities
    
    Now let $\mathbf{F} = (\mathbf{F_1}, \ldots, \mathbf{F_{n_r}}) ^ T$ be the $(n_r \times n_c)$ matrix of normalised fluorescence vectors of each round $r$ of a spot $s$. By independence between rounds, the probability of observing the fluorescence $\mathbf{f}$ from a spot of gene $g$ is just the product of the probabilities that each round $r$ is assigned to dye $d(g, r)$. In equations, this simplifies nicely to:
    
    $$ 
    \begin{aligned}
    \mathbb{P}[\mathbf{F} = \mathbf{f} \mid G = g] &= \prod_r \mathbb{P}[\mathbf{F_{r}} = \mathbf{f_r} \mid D = d(g, r)] \\ 
    &= \prod_r M_{\kappa} \exp \left( \kappa\mathbf{f_r} \cdot \mathbf{B_{d(g, r)}} \right) \\
    &= M_{\kappa}^{n_r} \exp \left( \kappa \sum_r \mathbf{f_r} \cdot \mathbf{B_{d(g, r)}} \right) \\ 
    &=  M_{\kappa}^{n_r} \exp(\kappa \mathbf{f \cdot b_g}), \end{aligned}
    $$
    
    where
    
    - $\mathbf{f} = (\mathbf{f_1}, \ldots, \mathbf{f_{n_r}}) ^ T$ is the observed round-normalised $(n_r \times n_c)$ fluorescence matrix of the spot,
    
      - $\mathbf{b_g} = (\mathbf{B_{d(g, 1)}}, \ldots, \mathbf{B_{d(g, n_r)}}) ^ T$ is the $(n_r \times n_c)$ matrix of the bled code for gene $g$,
    
      - the dot product $\mathbf{f \cdot b_g}$ is the [Frobenius Inner Product for Matrices](https://en.wikipedia.org/wiki/Frobenius_inner_product), ie: the sum of the elementwise product of the two matrices.
    
    We have so far only defined the probability of $\mathbf{F} = \mathbf{f}$ given $G = g$. We can find the probability of $G = g$ given $\mathbf{F} = \mathbf{f}$ using Bayes' Rule:
    
    $$ 
    \mathbb{P}[G = g \mid \mathbf{F} = \mathbf{f}] 
    = \dfrac{\mathbb{P}[\mathbf{F} = \mathbf{f} \mid G = g] \mathbb{P}[G = g]}{ \mathbb{P}[\mathbf{F} = \mathbf{f}]}. 
    $$
    
    For the priors, we will assume that:
    
    - $\mathbb{P}[G = g] = \frac{1}{n_g}$ (ie: all genes are equally likely),
    
      - $\mathbb{P}[\mathbf{F} = \mathbf{f}] = \sum_g \mathbb{P}[\mathbf{F} = \mathbf{f} \mid G = g] \mathbb{P}[G = g] = \frac{1}{n_g} \sum_g M_{\kappa}^{n_r} \exp(\kappa \mathbf{b}_g \cdot \mathbf{f})$, (ie: $\mathbf{f}$ comes from one of the genes)
    
    This gives us the final probability:
    
    $$ \mathbb{P}[G = g \mid \mathbf{F} = \mathbf{f}] = \frac{\exp(\kappa \mathbf{b}_g \cdot \mathbf{f})}{\sum_g \exp(\kappa\mathbf{b}_g \cdot \mathbf{f} )}$$


### 2: Bleed Matrix Calculation
The purpose of this step is to compute an updated estimate of the bleed matrix. 

Set some probability threshold $\gamma$ (in the config file $\gamma$ is called `gene_prob_thresold` and has default value 0.9). We define the following sets:

$$ 
\mathcal{S} = \{ s : p(s) \geq \gamma \},
$$

$$ 
G_{dr} = \{ g : d(g,r) = d \},
$$

$$ 
J_{dr} = \{ \mathbf{F_{sr}} : s \in \mathcal{S}, \ g_s \in G_{dr} \}
$$

In words, these can be described as follows:

- $\mathcal{S}$ is the set of spots with $s$ with probability $p(s) > \gamma$, ie: the high probability spots,

- $G_{dr}$ is the set of genes with dye $d$ in round $r$,

- $J_{dr}$ is the set of colours of high probability spots assigned to genes with dye $d$ in round $r$.

By taking the union of $J_{dr}$ across rounds, we end up with a set of reliable colour vector estimates for dye $d$:

$$ 
\mathcal{J}_d = \bigcup_r J_{dr}
$$


??? info "Why do we find spots like this?"

    The simpler way to find represantitive colours for each dye would be to look at the colours for all spots $s$ where $\mathbf{F_{sr}} \cdot \mathbf{B_{raw, d}}$ is above some threshold. This would give us a set of colours which are likely to be from dye $d$. 

    Our method is better for two reasons:
    
    1. The raw bleed matrix $\mathbf{B_{raw}}$ is not always good estimate of the bleed matrix.
    2. The central dyes have very similar colour spectra, so it is difficult to classify which dye the vector comes from by looking at round $r$ alone. By using information from adjacent rounds, we can more confidently ensure that the colours we are looking at are from dye $d$.

Let $\mathbf{J}$ be the $(n_{\text{good spots}} \times n_c)$ matrix form of the set $\mathcal{J}_d$. This just means each row of $\mathbf{J}$ corresponds to a good spot and each column corresponds to a channel. Compute the first [singular vectors](https://en.wikipedia.org/wiki/Singular_value_decomposition) of $\mathbf{J}$, ie: the optimal unit vectors $\boldsymbol{\omega} \in \mathbb{R}^{n_c}$ and $\boldsymbol{\eta} \in \mathbb{R}^{n_{\text{good spots}}}$ such that

$$ J_{s, c} \approx \lambda \eta_s \omega_c, $$

for some scalar $\lambda$. We then set $\mathbf{B_d} = \boldsymbol{\omega}$, which is a normalised fluorescence vector for dye $d.$

ADD IMAGE OF THIS PROCESS.

### 3: Free Bled Code Estimation

Our method of estimating $\mathbf{E_{g}}$ (and its tile dependent equivalent $\mathbf{D_{g,t}}$) should satisfy the following properties:

- If we have no spots, we should use a prior vector $\mathbf{E_g} = (\mathbf{B_{d(g, 1)}}, \ldots, \mathbf{B_{d(g, n_r)}}) ^ T$,

- We should allow each round $\mathbf{E_{g, r}}$ to scale $\mathbf{B_{d(g, r)}}$ easily,

- We should allow each round $\mathbf{E_{g, r}}$ to change the direction of $\mathbf{B_{d(g, r)}}$ less easily, but still allow it to change.

??? note "Why these properties?"

    Property 1 is necessary because for large gene panels we often have very few reads of each gene, meaning that we have very few samples to compute $\mathbf{E_g}$ and even fewer to compute $\mathbf{D_{g, t}}$.

    Property 2 is necessary because, as mentioned previously, different concentrations of bridge probes lead to systematic differences in brightness between genes in different rounds. We want to allow the brightness of each gene in each round to be scaled up or down without needing very many samples to do so.

    Property 3 is necessary because sometimes the way that a particular dye is expressed varies from gene to gene. An example of this is when the dyes are not completely washed out between rounds, leading to a small amount of bleedthrough from the previous round. This is shown below for a single gene.


Given $n$ round fluorescence vectors $\mathbf{f_1}, \ldots, \mathbf{f_n}$ and a prior unit vector $\mathbf{b}$, all in $\mathbb{R}^{n_c}$, we define the _parallel bayes mean_ of these as 


$$
\bar{\mathbf{f}}_{\alpha, \beta} = \dfrac{\alpha^2}{n + \alpha^2} \mathbf{b} + \dfrac{1}{n + \alpha^2} \bigg( \sum_i \mathbf{f_i \cdot b} \bigg) \mathbf{b} + 
\dfrac{1}{n+\beta^2} \sum_i \bigg( \mathbf{f_i} - (\mathbf{f_i} \cdot \mathbf{b})\mathbf{b} \bigg).
$$

The values $\alpha^2$ and $\beta^2$ are in the config file as `concentration_parameter_parallel` (default value 10) and `concentration_parameter_perpendicular` (default value 50) respectively.

!!! tip "How to choose and interpret $\alpha$ and $\beta$?"
    
    The formula for $\bar{\mathbf{f}}_{\alpha, \beta}$ is quite complcated, but it is actually quite easy to interpret once we understand what it is doing for different values of $\alpha$ and $\beta$.

      - If $\alpha = \beta = 0$, this is just the average. ie: $\bar{\mathbf{f}}_{0, 0} = \frac{1}{n} \sum_i \mathbf{f_i}$,
    
      - if $\alpha = \beta = m$, for some positive integer $m$, this is the average of $\mathbf{f_1}, \ldots, \mathbf{f_n}$ and $m$ copies of $\mathbf{b}$, ie: $\bar{\mathbf{f}}_{m, m} = \frac{1}{n + m} \sum_i \mathbf{f_i} + \frac{m}{n + m} \mathbf{b}$.
    
      - The component of $\bar{\mathbf{f}}_{\alpha, \beta}$ parallel to $\mathbf{b}$ has magnitude $\frac{\alpha^2 + \sum_i \mathbf{f_i} \cdot \mathbf{b}}{n + \alpha^2}$, which means that:
    
          - If $n << \alpha^2$ this magnitude is approximately 1, so this component is approximately $\mathbf{b}$,
      
          - If $n >> \alpha^2$ this magnitude is approximately $\frac{1}{n} \sum_i (\mathbf{f_i} \cdot \mathbf{b} ) \mathbf{b}$, which is the magnitude of the average $\mathbf{\bar{f}}$ in the direction $\mathbf{b}$.
    
      - The component of $\bar{\mathbf{f}}_{\alpha, \beta}$ perpendicular to $\mathbf{b}$ has magnitude $\frac{1}{n + \beta^2} \sum_i ( \mathbf{f_i} - (\mathbf{f_i} \cdot \mathbf{b})\mathbf{b} )$ which means that:
    
          - If $n << \beta^2$ this magnitude is approximately 0, so this component is approximately $\mathbf{0}$,
      
          - If $n >> \beta^2$ this magnitude is approximately $\frac{1}{n} \sum_i \mathbf{f_i} - (\mathbf{f_i} \cdot \mathbf{b})\mathbf{b}$, which is the average $\mathbf{\bar{f}}$ perpendicular to $\mathbf{b}$.
    
    From this analysis, we see that $\alpha^2$ is roughly the number of spots needed to scale $\mathbf{\bar{f}}_{\alpha, \beta}$ in the direction of $\mathbf{b}$, and $\beta^2$ is roughly the number of spots needed to scale $\mathbf{\bar{f}}_{\alpha, \beta}$ perpendicular to $\mathbf{b}$. 
    
    This is why we to set $\alpha^2 << \beta^2$. We want to easily scale the average in the direction of the prior vector, but not easily change its direction.


??? info "Derivation of the Parallel Bayes Mean"
    
    The formula for the parallel bayes mean is a [maximum a posteriori estimate](https://en.wikipedia.org/wiki/Maximum_a_posteriori_estimation) $\boldsymbol{\hat{\mu}}$ of the mean $\boldsymbol{\mu}$ of a normal distribution with a normal prior $\boldsymbol{\mu} \sim \mathcal{N}(\mathbf{b}, \Sigma)$, where $\Sigma$ is a diagonal matrix with $1/\alpha^2$ in the direction of $\mathbf{b}$ and $1/\beta^2$ in the direction orthogonal to $\mathbf{b}$.
    
    Let $\mathbf{F_1}, \ldots, \mathbf{F_n}$  be the round $r$ fluorescence vectors of spots assigned to gene $g$ with high probability and let $\mathbf{B}_{d(g,r)}$ be the prior unit vector.
    
    To begin, assume the vectors $\mathbf{F_1}, \ldots, \mathbf{F_n}$ are i.i.d normal random variables with mean $\boldsymbol{\mu}$ and covariance $I_{n_c}$, wihch means the sample mean is also normal with mean $\boldsymbol{\mu}$ and covariance $\frac{I_{n_c}}{n}$. Impose a normal prior on the space of possible means:
    
    $$
    \overline{\mathbf{F}} \sim \mathcal{N} \bigg( \boldsymbol{\mu}, \frac{\boldsymbol{I_{n_c}}}{n} \bigg)
    $$
    
    $$
    \boldsymbol{\mu} \sim \mathcal{N}(\mathbf{B}_{d(g,r)}, \Sigma)
    $$
    
    where 
    
    $$
    \Sigma = \text{Diag}\left(\frac{1}{\alpha^2}, \frac{1}{\beta^2}, \ldots, \frac{1}{\beta^2}\right),
    $$
    
    in the orthonormal basis $\mathbf{v}_1 = \mathbf{B}_{d(g,r)}$, and everything else orthogonal to this.

    Set $\boldsymbol{\Lambda} =\boldsymbol{\Sigma}^{-1}$, $\mathbf{b} = \mathbf{B_{d(g,r)}}$, and recall that the normal is a conjugate prior, meaning the posterior $\boldsymbol{\mu} \mid \mathbf{\overline{F}}$ is also normal. 
    
    To find its mode we'll solve for the zeros of the derivative of its log-density. The log-density of $\boldsymbol{\mu} \mid \mathbf{\overline{F}}$ is given by
    
    $$
    \begin{aligned}
    l(\boldsymbol{\mu}) &= \log P(\boldsymbol{\mu}| \overline{\mathbf{F}} = \mathbf{f}) \\ \\
           &= \log P(\boldsymbol{\mu}) + \log P(\overline{\mathbf{F}} = \mathbf{f} | \boldsymbol{\mu}) + C \\ \\
           &= -\frac{1}{2} (\boldsymbol{\mu} - \mathbf{b})^T \boldsymbol{\Lambda} (\boldsymbol{\mu} - \mathbf{b}) - \frac{n}{2} (\boldsymbol{\mu} - \mathbf{f})^T (\boldsymbol{\mu} - \mathbf{f}) + C
    \end{aligned}
    $$
    
    This has derivative
    
    $$
    \frac{\partial l}{\partial \boldsymbol{\mu}} = - \boldsymbol{\Lambda} (\boldsymbol{\mu} - \boldsymbol{b}) - n(\boldsymbol{\mu} - \mathbf{f})
    $$
    
    Setting this to $\mathbf{0}$, rearranging for $\boldsymbol{\mu}$ and using the fact that
    
    $$
    \boldsymbol{\Lambda} \mathbf{v} = 
    \begin{cases}
    \alpha^2 \mathbf{v} & \text{if } \mathbf{v} = \lambda\mathbf{b} \\ \\
    \beta^2 \mathbf{v} & \text{otherwise}
    \end{cases}
    $$
    
    we get
    
    $$
    \begin{aligned}
    \boldsymbol{\hat{\mu}} &= (\Lambda + nI)^{-1}(\Lambda \mathbf{b} + n\mathbf{f}) \\ \\ 
        &= (\Lambda + nI)^{-1}(\alpha^2 \mathbf{b} + n\mathbf{f}) \\ \\
        &= (\Lambda + nI)^{-1}(\alpha^2 \mathbf{b} + n(\mathbf{f} \cdot \mathbf{b})\mathbf{b} + n(\mathbf{f} - (\mathbf{f} \cdot \mathbf{b})\mathbf{b})) \\ \\
        &= (\Lambda + nI)^{-1}((\alpha^2 + n \mathbf{f} \cdot \mathbf{b})\mathbf{b} + n(\mathbf{f} - (\mathbf{f} \cdot \mathbf{b})\mathbf{b}))\\ \\
    &= \dfrac{(\alpha^2 + n \mathbf{f} \cdot \mathbf{b})}{n + \alpha^2} \mathbf{b} +
     \dfrac{n}{n+\beta^2} \bigg( \mathbf{f} - (\mathbf{f} \cdot \mathbf{b})\mathbf{b} \bigg)
    \end{aligned}
    $$
    
    Plugging in the observed sample mean $\mathbf{f} = \frac{1}{n}\sum_i \mathbf{f_{i, r}}$ yields our estimate $\boldsymbol{\hat{\mu}}$.


### 4: Round and Channel Normalisation

The purpose of this step is to find a scale factor $V_{rc}$ for each round $r$ and channel $c$ which gets as many of our spots as close as possible to their target values. Since $E_{grc}$ is a representative sample of the colour spectrum of gene $g$ in round $r$ and channel $c$, we want the scale factor to satisfy 

$$
E_{grc}V_{rc} \approx T_{c},
$$

where $T_c$ is the target value for channel $c$ when 

To begin, fix a round $r$ and channel $c$ and let $d_{max}(c)$ be the dye which is most intense in channel $c$. We define a target value $T_d$ for each dye $d$ in its maximal channel $c_{max}(d)$. Now let $G_{r,d}$ be the set of genes with dye $d$ in round $r$, and define the loss function 

$$
L(V_{r, c}) = \sum_{g \in G_{r, \ d_{max}(c)}} \sqrt{N_{g}} \  \bigg( V_{r, c} \ E_{grc} - T_{d_{max}(c)} \bigg)^2,
$$ 

where $N_g$ is the number of high probability spots assigned to gene $g$. There is no reason this has to be a square root, though if it is not, too much influence is given to the most frequent genes. Minimise this loss to obtain

$$
V_{r, c} = \dfrac{ \sum_{g \in G_{r, \ d_{max}(c) }} \sqrt{N_g} E_{grc} T_{d_{max}(c)} } { \sum_{g \in G_{r, \ d_{max}(c) }} \sqrt{N_g} E_{grc}^2 },
$$

which is our optimal value!

Now define the _constrained bled codes_, which we will just call _bled codes_ 

$$
K_{grc} = E_{grc}V_{rc}.
$$

We will use these instead of $E_{grc}$ from here onwards.

### 5: Tile Normalisation

The purpose of this step is to remove brightness differences between tiles, and improve the round and channel normalisation we found in the previous step. We do this by finding a scale factor $Q_{trc}$ such that 

$$
Q_{trc} D_{gtrc} \approx K_{grc},
$$

where $\mathbf{D_{g, t,}}$ is the tile-dependent bled code for gene $g$ in tile $t$ defined in step 3 and $\mathbf{K_g}$ is the constrained bled code for gene $g$ defined in step 4.

Our method works in a similar way to step 4: fix a tile $t$, round $r$ and channel $c$ and as above, let $G_{r,d}$ be the genes with dye $d$ in round $r$. Define the loss

$$
L(Q_{trc}) = \sum_{g \in G_{r, \ d_{max}(c)}} \sqrt{N_{g,t}} \  \bigg( Q_{trc} \ D_{gtrc} - K_{grc} \bigg)^2,
$$ 

where $N_{g, t}$ is the number of high probability spots of gene $g$ in tile $t$. Remember that $K_{grc} = E_{grc}V_{r, c},$ so writing this in full yields

$$
L(Q_{trc}) = \sum_{g \in G_{r, \ d_{max}(c)}} \sqrt{N_{gt}} \  \bigg( Q_{trc} \ D_{gtrc} - E_{grc}V_{rc} \bigg)^2.
$$ 

This means that if $D_{gtrc} \approx E_{grc}$ for all genes $g$ then $Q_{t,r,c} \approx V_{r,c}$. This means that $\mathbf{Q}$ is correcting for tile differences. Then why does it have indices for $r$ and $c$? Because the way that the brightness varies between tiles may be completely independent for different round-channel pairs. This is addressed further in the diagnostics. Minimising this loss yields:

$$
Q_{trc} = \dfrac{ \sum_{g \in G_{r, \ d_{max}(c) }} \sqrt{N_{gt}} \ K_{grc}  D_{gtrc}} { \sum_{g \in G_{r,\ d_{max}(c) }} \sqrt{N_{gt}}  D_{gtrc}^2 }.
$$

<figure markdown="span">
  ![Image title](images/algorithm/call_spots/tile scale regression.png)
</figure>
*We can use the `view_tile_scale_regression` diagnostic to view the regression for a single tile. Note that the regressions seem to have a high $r^2$ value and the slopes are significantly different even within channels.*

<figure markdown="span">
  ![Image title](images/algorithm/call_spots/different scales.png)
</figure>
*We can use the `view_scale_factors` diagnostic to view the round/channel scale, the tile scale and the tile scales relative to the round/channel scales. The final plot being non-uniform implies that the tile scale corrections are independent between different $(r, c)$ pairs.*

### 6 and 7: Application of Scales, Computation of Final Scores and Bleed Matrix

The purpose of this step is to bring all the components together and compute the final scores and bleed matrix.

Now that we have the tile scale $Q_{t,r,c}$, we multiply it by the initial scale factor $P_{trc}$ to get our final colour normalisation factor $A_{trc}$:

$$
A_{trc} = P_{trc} Q_{trc}.
$$

This is important as all our calculations have been done on preprocessed spot colours which have already been multiplied by $\mathbf{P}$. We apply this scale to all of our spot colours $F$ by pointwise multiplication.

Next, we compute the final probabilities and dot product scores. We might ask at this point whether we should use:

1. The tile-independent free bled codes $E_{grc}$, 

2. the tile-dependent free bled codes $D_{gtrc}$ or,

3. the constrained bled codes $K_{grc}$? 

The best answer is definitely the constrained bled codes $K_{grc}$! We calculated $\mathbf{E}$ and $\mathbf{D}$ as important summary statistics which act as representative samples for each gene in the case of $\mathbf{E}$, and for each gene and tile in the case of $\mathbf{D}$. These were only computed to facilitate the computations of $\mathbf{A}$ and $\mathbf{K}$.

While it may seem more accurate to have a different gene code for each tile, the estimates $\mathbf{D_{g, t}}$ are noisy due to small numbers of samples. The colour normalisation factor $\mathbf{A}$ was calculated specifically to maximise similarity of the **tile-dependent** free codes $\mathbf{D}$ with the **tile-independent** constrained codes $\mathbf{K}$, meaning that multiplying spot colours by $\mathbf{A}$ has the dual effect of

1. homogenising them across tiles and,

2. bringing them all close to the constrained codes $\mathbf{K}$.

With that in mind, we compute:

1. Final gene probabilities using the scaled spots $\mathbf{AF}$ and comparing against the constrained codes $\mathbf{K}$

2. Final Dot Products using the scaled spots $\mathbf{AF}$ and comparing against the constrained codes $\mathbf{K}$. These would not have been accurate in step 1 as we had no model of how each gene varied in brightness between rounds, but now this is something we have accounted for in $\mathbf{K}$.

3. The Final Bleed Matrix using the same method as discussed in step 2, but with updated gene probabilities.
